mesonbuild/mintro.py

```diff
@@ -298,48 +304,103 @@ def list_projinfo_from_source(sourcedir):
-    except FileNotFoundError:
-        installdata = None
-
-    if options.list_targets:
-        list_targets(coredata, builddata, installdata)
-    elif options.list_installed:
-        list_installed(installdata)
-    elif options.target_files is not None:
-        list_target_files(options.target_files, coredata, builddata)
-    elif options.buildsystem_files:
-        list_buildsystem_files(builddata)
-    elif options.buildoptions:
-        list_buildoptions(coredata, builddata)
-    elif options.tests:
-        list_tests(testdata)
-    elif options.benchmarks:
-        list_tests(benchmarkdata)
-    elif options.dependencies:
-        list_deps(coredata)
-    elif options.projectinfo:
-        list_projinfo(builddata)
-    else:
+    # Load build data to make sure that the version matches
+    # TODO Find a better solution for this
+    cdata.load(options.builddir)
+
+    results = []
+    toextract = []
+
+    if options.all or options.benchmarks:
+        toextract += ['benchmarks']
+    if options.all or options.buildoptions:
+        toextract += ['buildoptions']
+    if options.all or options.buildsystem_files:
+        toextract += ['buildsystem_files']
+    if options.all or options.dependencies:
+        toextract += ['dependencies']
+    if options.all or options.list_installed:
+        toextract += ['installed']
+    if options.all or options.projectinfo:
+        toextract += ['projectinfo']
+    if options.all or options.list_targets:
+        toextract += ['targets']
+    if options.target_files is not None:
+        targets_file = os.path.join(infodir, 'intro-targets.json')
+        with open(targets_file, 'r') as fp:
+            targets = json.load(fp)
+        builddata = build.load(options.builddir) # TODO remove this in a breaking changes PR
+        results += [list_target_files(options.target_files, targets, builddata)]
+    if options.all or options.tests:
+        toextract += ['tests']
+
+    for i in toextract:
+        curr = os.path.join(infodir, 'intro-{}.json'.format(i))
+        if not os.path.isfile(curr):
+            print('Introspection file {} does not exist.'.format(curr))
+            return 1
+        with open(curr, 'r') as fp:
+            results += [(i, json.load(fp))]
+
+    indent = options.indent if options.indent > 0 else None
+
+    if len(results) == 0 and not options.force_dict:
         print('No command specified')
         return 1
+    elif len(results) == 1 and not options.force_dict:
+        # Make to keep the existing output format for a single option
+        print(json.dumps(results[0][1], indent=indent))
+    else:
+        out = {}
+        for i in results:
+            out[i[0]] = i[1]
+        print(json.dumps(out, indent=indent))
     return 0
+
+def write_intro_info(intro_info, info_dir):
+    for i in intro_info:
+        out_file = os.path.join(info_dir, 'intro-{}.json'.format(i[0]))
+        with open(out_file, 'w') as fp:
+            json.dump(i[1], fp)
```

Comment:
Do we want to write this atomically? That would help IDEs to get the complete json without timeouts etc.